receiver/awscontainerinsightreceiver/internal/efa/efaSysfs.go

```diff
@@ -0,0 +1,443 @@
+			metricName:    metricName,
+		},
+	}
+	deltaValue, found := s.deltaCalculator.Calculate(key, value, timestamp)
+	if found {
+		metric.AddField(metricName, deltaValue)
+	}
+}
+
+func (s *Scraper) init() {
+}
+
+func (s *Scraper) startScrape(ctx context.Context) {
+	ticker := time.NewTicker(s.collectionInterval)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			err := s.scrape()
+			if err != nil {
+				s.logger.Warn("Failed to scrape EFA metrics from filesystem", zap.Error(err))
+			}
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+func (s *Scraper) scrape() error {
+	exists, err := s.sysFsReader.EfaDataExists()
+	if err != nil {
+		return err
+	}
+	if !exists {
+		return nil
+	}
+
+	timestamp := time.Now()
+
+	devices, err := s.parseEfaDevices()
+	if err != nil {
+		return err
+	}
+
+	s.store = &efaStore{
+		timestamp: timestamp,
+		devices:   devices,
+	}
+
+	return nil
+}
+
+func (s *Scraper) parseEfaDevices() (*efaDevices, error) {
+	deviceNames, err := s.sysFsReader.ListDevices()
+	if err != nil {
+		return nil, err
+	}
+
+	devices := make(efaDevices, len(deviceNames))
+	for _, name := range deviceNames {
+		counters, err := s.parseEfaDevice(name)
+		if err != nil {
+			return nil, err
+		}
+
+		devices[name] = counters
+	}
+
+	return &devices, nil
+}
+
+func (s *Scraper) parseEfaDevice(deviceName efaDeviceName) (*efaCounters, error) {
+	ports, err := s.sysFsReader.ListPorts(deviceName)
+	if err != nil {
+		return nil, err
+	}
+
+	counters := new(efaCounters)
+	for _, port := range ports {
+		err := s.readCounters(deviceName, port, counters)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return counters, nil
+}
+
+func (s *Scraper) readCounters(deviceName efaDeviceName, port string, counters *efaCounters) error {
+	var errs error
+	reader := func(counter string) uint64 {
+		value, err := s.sysFsReader.ReadCounter(deviceName, port, counter)
+		if err != nil {
+			errs = errors.Join(errs, err)
+			return 0
+		}
+		return value
+	}
+
+	for counter := range counterNames {
+		switch counter {
+		case counterRdmaReadBytes:
+			counters.rdmaReadBytes += reader(counter)
+		case counterRdmaWriteBytes:
+			counters.rdmaWriteBytes += reader(counter)
+		case counterRdmaWriteRecvBytes:
+			counters.rdmaWriteRecvBytes += reader(counter)
+		case counterRxBytes:
+			counters.rxBytes += reader(counter)
+		case counterRxDrops:
+			counters.rxDrops += reader(counter)
+		case counterTxBytes:
+			counters.txBytes += reader(counter)
+		}
+	}
+
+	return errs
+}
+
+func defaultSysFsReader(logger *zap.Logger) sysFsReader {
+	return &sysfsReaderImpl{
+		logger: logger,
+	}
+}
+
+type sysfsReaderImpl struct {
+	logger *zap.Logger
+}
+
+var _ sysFsReader = (*sysfsReaderImpl)(nil)
+
+func (r *sysfsReaderImpl) EfaDataExists() (bool, error) {
+	info, err := os.Stat(efaPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	stat, ok := info.Sys().(*syscall.Stat_t)
+	if !ok {
+		return true, nil
```

Comment:
This would essentially mean that we are not running on Linux.  But you're right, we should skip in that case so we remember to implement this logic for Windows if we ever add support for that.  Changed in next commit.